### ISSUE TITLE
Weird crash of git rev-parse ... rm: cannot remove '/home/yoh/.cache/git-annex/locales/...

### ISSUE THREAD

**Issue by yarikoptic** [datalad/datalad#4976]
I decided to troubleshoot (log at higher level etc) "why creating subdatasets is so slow when done by `addurls`".  And quickly it has crashed with

```
  File "/mnt/scrap/tmp/abcd/datalad/datalad/core/local/save.py", line 304, in __call__
    if start_commit != pds_repo.get_hexsha()
  File "/mnt/scrap/tmp/abcd/datalad/datalad/support/gitrepo.py", line 1808, in get_hexsha
    return self.call_git_oneline(cmd)
  File "/mnt/scrap/tmp/abcd/datalad/datalad/support/gitrepo.py", line 2290, in call_git_oneline
    .format(["git"] + args, lines))
AssertionError: Expected ['git', 'rev-parse', '--quiet', '--verify', 'HEAD^{commit}'] to return single line, but it returned ["rm: cannot remove '/home/yoh/.cache/git-annex/locales/bb375eb6ec0d2706f1723307f068911a': Directory not empty", '01a0149f261576dbaf2823d63ac0a789a02af1df']
```

<details>
<summary>full trace with invocation etc</summary> 

```shell
smaug:/mnt/scrap/tmp/abcd
$> DATALAD_LOG_TIMESTAMP=1 DATALAD_LOG_NAME=1 DATALAD_LOG_LEVEL=4 DATALAD_LOG_PID=1 DATALAD_LOG_TARGET=$PWD/datalad-full4-part-`mdate`-1.log  datalad-nda/scripts/datalad-nda --pdb add2datalad -i <(xzcat ~/proj/abcd/datastructure_manifest.txt.xz) -d testds-fast-full4-part --fast
2020-10-02 09:04:34,643 [INFO   ] Creating a new dataset at /mnt/scrap/tmp/abcd/testds-fast-full4-part
2020-10-02 09:04:39,574 [INFO   ] Creating a helper procedure
initremote datalad ok
(recording state in git...)
2020-10-02 09:04:42,230 [INFO   ] Reading entire file
2020-10-02 09:04:47,199 [INFO   ] Read 4166278 lines
 81%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▏                                 | 3367150/4166278 [00:18<00:04, 174639.02it/s]2020-10-02 09:05:10,681 [INFO   ] Loaded 4166276 records from 40 submissions for 40 datasets.
initremote datalad ok
(recording state in git...)
2020-10-02 09:05:28,210 [INFO   ] Processing for submission 21948
2020-10-02 09:05:28,897 [INFO   ] Getting records only for submission_id=21948
2020-10-02 09:05:30,036 [INFO   ] Got 331758 records
2020-10-02 09:05:32,469 [INFO   ] Processed entire file: got 331758 files, 1472 subdatasets with 7 files having multiple URLs to possibly reach them
2020-10-02 09:05:32,470 [WARNING]  - dataset_description.json: 45848 records
|   - README: 45848 records
|   - CHANGES: 45848 records
|   - task-nback_bold.json: 1197 records
|   - task-MID_bold.json: 1235 records
|   - task-SST_bold.json: 1191 records
|   - task-rest_bold.json: 1470 records
2020-10-02 09:05:45,461 [INFO   ] Saved output to <_io.TextIOWrapper name='/mnt/scrap/tmp/abcd/testds-fast-full4-part/sourcedata/submission-21948.csv' mode='w' encoding='UTF-8'>
initremote datalad ok
(recording state in git...)
initremote datalad ok
(recording state in git...)
Traceback (most recent call last):
  File "datalad-nda/scripts/datalad-nda", line 424, in <module>
    main()
  File "/usr/lib/python3/dist-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/usr/lib/python3/dist-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/lib/python3/dist-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/lib/python3/dist-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "datalad-nda/scripts/datalad-nda", line 244, in add2datalad
    on_failure="stop",
  File "/mnt/scrap/tmp/abcd/datalad/datalad/distribution/dataset.py", line 503, in apply_func
    return f(**kwargs)
  File "/mnt/scrap/tmp/abcd/datalad/datalad/interface/utils.py", line 481, in eval_func
    return return_func(generator_func)(*args, **kwargs)
  File "/mnt/scrap/tmp/abcd/datalad/datalad/interface/utils.py", line 469, in return_func
    results = list(results)
  File "/mnt/scrap/tmp/abcd/datalad/datalad/interface/utils.py", line 400, in generator_func
    allkwargs):
  File "/mnt/scrap/tmp/abcd/datalad/datalad/interface/utils.py", line 539, in _process_results
    for res in results:
  File "/mnt/scrap/tmp/abcd/datalad/datalad/plugin/addurls.py", line 897, in __call__
    return_type='generator')
  File "/mnt/scrap/tmp/abcd/datalad/datalad/interface/utils.py", line 400, in generator_func
    allkwargs):
  File "/mnt/scrap/tmp/abcd/datalad/datalad/interface/utils.py", line 539, in _process_results
    for res in results:
  File "/mnt/scrap/tmp/abcd/datalad/datalad/core/local/create.py", line 540, in __call__
    path=tbds.path,
  File "/mnt/scrap/tmp/abcd/datalad/datalad/distribution/dataset.py", line 503, in apply_func
    return f(**kwargs)
  File "/mnt/scrap/tmp/abcd/datalad/datalad/interface/utils.py", line 481, in eval_func
    return return_func(generator_func)(*args, **kwargs)
  File "/mnt/scrap/tmp/abcd/datalad/datalad/interface/utils.py", line 469, in return_func
    results = list(results)
  File "/mnt/scrap/tmp/abcd/datalad/datalad/interface/utils.py", line 400, in generator_func
    allkwargs):
  File "/mnt/scrap/tmp/abcd/datalad/datalad/interface/utils.py", line 539, in _process_results
    for res in results:
  File "/mnt/scrap/tmp/abcd/datalad/datalad/core/local/save.py", line 304, in __call__
    if start_commit != pds_repo.get_hexsha()
  File "/mnt/scrap/tmp/abcd/datalad/datalad/support/gitrepo.py", line 1808, in get_hexsha
    return self.call_git_oneline(cmd)
  File "/mnt/scrap/tmp/abcd/datalad/datalad/support/gitrepo.py", line 2290, in call_git_oneline
    .format(["git"] + args, lines))
AssertionError: Expected ['git', 'rev-parse', '--quiet', '--verify', 'HEAD^{commit}'] to return single line, but it returned ["rm: cannot remove '/home/yoh/.cache/git-annex/locales/bb375eb6ec0d2706f1723307f068911a': Directory not empty", '01a0149f261576dbaf2823d63ac0a789a02af1df']

> /mnt/scrap/tmp/abcd/datalad/datalad/support/gitrepo.py(2290)call_git_oneline()
-> .format(["git"] + args, lines))
```
</details>

full log dump to that point is on smaug:/mnt/scrap/tmp/datalad-full4-part-20201002-1.log

git-annex is some recent snapshot...  may be that cache/locales is smth contributing to the slowness... (inspired by this -- initiated https://github.com/datalad/datalad-extensions/issues/36)

edit 1: slowness is probably largely due to sandwiching too much of datalad!! It looks like `datalad-nda python script -> bash procedure -> git annex initremote -> git-annex-remote-datalad -> git/git-annex` if I got it right...

<details>
<summary>WTF</summary> 

# WTF
## datalad 
  - full_version: 0.13.3.dev351-g2e67
  - version: 0.13.3.dev351
## dependencies 
  - annexremote: 1.4.3
  - appdirs: 1.4.3
  - boto: 2.44.0
  - cmd:7z: 16.02
  - cmd:annex: 8.20200908-gcfc74c2f4
  - cmd:bundled-git: 2.28.0
  - cmd:git: 2.28.0
  - cmd:system-git: 2.28.0
  - cmd:system-ssh: 7.9p1
  - exifread: 2.1.2
  - git: 3.0.5
  - gitdb: 2.0.5
  - humanize: 0.5.1
  - iso8601: 0.1.11
  - keyring: 17.1.1
  - keyrings.alt: 3.1.1
  - msgpack: 0.5.6
  - mutagen: 1.40.0
  - requests: 2.21.0
  - wrapt: 1.10.11
## environment 
  - GIT_PAGER: less --no-init --quit-if-one-screen
  - LANG: en_US
  - LANGUAGE: en_US:en
  - LC_ADDRESS: en_US.UTF-8
  - LC_COLLATE: en_US.UTF-8
  - LC_CTYPE: en_US.UTF-8
  - LC_IDENTIFICATION: en_US.UTF-8
  - LC_MEASUREMENT: en_US.UTF-8
  - LC_MESSAGES: en_US.UTF-8
  - LC_MONETARY: en_US.UTF-8
  - LC_NAME: en_US.UTF-8
  - LC_NUMERIC: en_US.UTF-8
  - LC_PAPER: en_US.UTF-8
  - LC_TELEPHONE: en_US.UTF-8
  - LC_TIME: en_US.UTF-8
  - PATH: /home/yoh/.tmp/ga-RIFzW89/git-annex.linux:/mnt/scrap/tmp/abcd/datalad/venvs/dev3/bin:/home/yoh/gocode/bin:/home/yoh/gocode/bin:/home/yoh/gocode/bin:/home/yoh/bin:/home/yoh/.local/bin:/usr/local/bin:/usr/bin:/bin:/usr/games:/sbin:/usr/sbin:/usr/local/sbin
## extensions 
  - container: 
    - description: Containerized environments
    - entrypoints: 
      - datalad_container.containers_add.ContainersAdd: 
        - class: ContainersAdd
        - load_error: None
        - module: datalad_container.containers_add
        - names: 
          - containers-add
          - containers_add
      - datalad_container.containers_list.ContainersList: 
        - class: ContainersList
        - load_error: None
        - module: datalad_container.containers_list
        - names: 
          - containers-list
          - containers_list
      - datalad_container.containers_remove.ContainersRemove: 
        - class: ContainersRemove
        - load_error: None
        - module: datalad_container.containers_remove
        - names: 
          - containers-remove
          - containers_remove
      - datalad_container.containers_run.ContainersRun: 
        - class: ContainersRun
        - load_error: None
        - module: datalad_container.containers_run
        - names: 
          - containers-run
          - containers_run
    - load_error: None
    - module: datalad_container
    - version: 1.0.1
## git-annex 
  - build flags: 
    - Assistant
    - Webapp
    - Pairing
    - Inotify
    - DBus
    - DesktopNotify
    - TorrentParser
    - MagicMime
    - Feeds
    - Testsuite
    - S3
    - WebDAV
  - dependency versions: 
    - aws-0.22
    - bloomfilter-2.0.1.0
    - cryptonite-0.26
    - DAV-1.3.4
    - feed-1.3.0.1
    - ghc-8.8.4
    - http-client-0.6.4.1
    - persistent-sqlite-2.10.6.2
    - torrent-10000.1.1
    - uuid-1.3.13
    - yesod-1.6.1.0
  - key/value backends: 
    - SHA256E
    - SHA256
    - SHA512E
    - SHA512
    - SHA224E
    - SHA224
    - SHA384E
    - SHA384
    - SHA3_256E
    - SHA3_256
    - SHA3_512E
    - SHA3_512
    - SHA3_224E
    - SHA3_224
    - SHA3_384E
    - SHA3_384
    - SKEIN256E
    - SKEIN256
    - SKEIN512E
    - SKEIN512
    - BLAKE2B256E
    - BLAKE2B256
    - BLAKE2B512E
    - BLAKE2B512
    - BLAKE2B160E
    - BLAKE2B160
    - BLAKE2B224E
    - BLAKE2B224
    - BLAKE2B384E
    - BLAKE2B384
    - BLAKE2BP512E
    - BLAKE2BP512
    - BLAKE2S256E
    - BLAKE2S256
    - BLAKE2S160E
    - BLAKE2S160
    - BLAKE2S224E
    - BLAKE2S224
    - BLAKE2SP256E
    - BLAKE2SP256
    - BLAKE2SP224E
    - BLAKE2SP224
    - SHA1E
    - SHA1
    - MD5E
    - MD5
    - WORM
    - URL
    - X*
  - operating system: linux x86_64
  - remote types: 
    - git
    - gcrypt
    - p2p
    - S3
    - bup
    - directory
    - rsync
    - web
    - bittorrent
    - webdav
    - adb
    - tahoe
    - glacier
    - ddar
    - git-lfs
    - httpalso
    - hook
    - external
  - supported repository versions: 
    - 8
  - upgrade supported from repository versions: 
    - 0
    - 1
    - 2
    - 3
    - 4
    - 5
    - 6
    - 7
  - version: 8.20200908-gcfc74c2f4
## location 
  - path: /mnt/scrap/tmp/abcd
  - type: directory
## metadata_extractors 
  - annex: 
    - load_error: None
    - module: datalad.metadata.extractors.annex
    - version: None
  - audio: 
    - load_error: None
    - module: datalad.metadata.extractors.audio
    - version: None
  - datacite: 
    - load_error: None
    - module: datalad.metadata.extractors.datacite
    - version: None
  - datalad_core: 
    - load_error: None
    - module: datalad.metadata.extractors.datalad_core
    - version: None
  - datalad_rfc822: 
    - load_error: None
    - module: datalad.metadata.extractors.datalad_rfc822
    - version: None
  - exif: 
    - load_error: None
    - module: datalad.metadata.extractors.exif
    - version: None
  - frictionless_datapackage: 
    - load_error: None
    - module: datalad.metadata.extractors.frictionless_datapackage
    - version: None
  - image: 
    - load_error: None
    - module: datalad.metadata.extractors.image
    - version: None
  - xmp: 
    - load_error: None
    - module: datalad.metadata.extractors.xmp
    - version: None
## python 
  - implementation: CPython
  - version: 3.7.3
## system 
  - distribution: debian/10.0
  - encoding: 
    - default: utf-8
    - filesystem: utf-8
    - locale.prefered: UTF-8
  - max_path_length: 275
  - name: Linux
  - release: 4.19.0-5-amd64
  - type: posix
  - version: #1 SMP Debian 4.19.37-5+deb10u1 (2019-07-19)

</details>

**Comment by kyleam** [comment]
Hmm, so given the extra lines on stdout, the response of `call_git_oneline` seems appropriate.  I'm trying to figure out where specifically the `rm ... -> stdout` is happening.

**Comment by yarikoptic** [comment]
FWIW: `rm` itself comes from annex'es `standalone/linux/skel/runshell` and that `runshell` is a shim used by every `git-annex` or bundled `git` call

<details>
<summary>grep</summary> 

```shell
$> grep runshell /usr/lib/git-annex.linux/git*(*)
/usr/lib/git-annex.linux/git:if [ ! -e "$base/runshell" ]; then
/usr/lib/git-annex.linux/git:	echo "** cannot find $base/runshell" >&2
/usr/lib/git-annex.linux/git:exec "$base/runshell" git "$@"
/usr/lib/git-annex.linux/git-annex:if [ ! -e "$base/runshell" ]; then
/usr/lib/git-annex.linux/git-annex:	echo "** cannot find $base/runshell" >&2
/usr/lib/git-annex.linux/git-annex:exec "$base/runshell" git-annex "$@"
/usr/lib/git-annex.linux/git-annex-shell:if [ ! -e "$base/runshell" ]; then
/usr/lib/git-annex.linux/git-annex-shell:	echo "** cannot find $base/runshell" >&2
/usr/lib/git-annex.linux/git-annex-shell:exec "$base/runshell" git-annex-shell "$@"
/usr/lib/git-annex.linux/git-annex-webapp:if [ ! -e "$base/runshell" ]; then
/usr/lib/git-annex.linux/git-annex-webapp:	echo "** cannot find $base/runshell" >&2
/usr/lib/git-annex.linux/git-annex-webapp:exec "$base/runshell" git-annex webapp "$@"
/usr/lib/git-annex.linux/git-receive-pack:if [ ! -e "$base/runshell" ]; then
/usr/lib/git-annex.linux/git-receive-pack:	echo "** cannot find $base/runshell" >&2
/usr/lib/git-annex.linux/git-receive-pack:exec "$base/runshell" git-receive-pack "$@"
/usr/lib/git-annex.linux/git-shell:if [ ! -e "$base/runshell" ]; then
/usr/lib/git-annex.linux/git-shell:	echo "** cannot find $base/runshell" >&2
/usr/lib/git-annex.linux/git-shell:exec "$base/runshell" git-shell "$@"
/usr/lib/git-annex.linux/git-upload-pack:if [ ! -e "$base/runshell" ]; then
/usr/lib/git-annex.linux/git-upload-pack:	echo "** cannot find $base/runshell" >&2
/usr/lib/git-annex.linux/git-upload-pack:exec "$base/runshell" git-upload-pack "$@"
```
</details>

**Comment by kyleam** [comment]
> `rm` itself comes from annex'es `standalone/linux/skel/runshell`

Thanks.  And it looks like that's redirecting to stdout:

```sh
	for localecache in $HOME/.cache/git-annex/locales/*; do
		cachebase=$(cat "$localecache/base" 2>/dev/null || true)
		if [ ! -d "$cachebase" ] || ! cmp "$localecache/buildid" "$cachebase/buildid" >/dev/null 2>&1 ; then
			rm -rf "$localecache" 2>&1 || true
		fi
	done
```

I wonder why.

**Comment by kyleam** [comment]
The stderr to stdout switch happened with a8a0f7fc5 (refuse to run if we can't write to HOME, 2018-07-10), but I'm not sure why.

**Comment by yarikoptic** [comment]
reported against git-annex: https://git-annex.branchable.com/bugs/standalone_runshell_can_race_and_fail_to_remove___96____126____47__.cache__47__git-annex__47__locales__47____96___dirs/

**Comment by kyleam** [comment]
> The stderr to stdout switch happened with a8a0f7fc5 (refuse to run if we can't write to HOME, 2018-07-10), but I'm not sure why.

Based on the response to the above report, it looks like the output was [supposed to go to /dev/null][1].  It now does as of cd9a60bc7 (runshell: Fix a edge case where rm errors were sent to stdout, which could confuse things parsing git-annex output, 2020-10-05).

[1]: http://git-annex.branchable.com/bugs/standalone_runshell_can_race_and_fail_to_remove___96____126____47__.cache__47__git-annex__47__locales__47____96___dirs/#comment-7c9ea9fc493c65fab2e389a25a1ccef2

**Comment by yarikoptic** [comment]
oh, may be I misattributed the fix to [8.20200908-163-gd74d97896](https://git.kitenet.net/index.cgi/git-annex.git/commit/?id=d74d9789688fde1151ece6a751d1c549af87de84) but will leave this link FTR and reference to annex version
